src/ssl.jl

```diff
@@ -643,18 +674,18 @@ function Base.close(ssl::SSLStream, shutdown::Bool=true)
     # if we've already finalized, no further action needed
     ssl.ssl.ssl == C_NULL && return
     # close operations
-    Base.@lock ssl.lock begin
+    # close underlying io
+    try
+        Base.close(ssl.io)
+    catch e
+        e isa Base.IOError || rethrow()
+    end
+    Base.@lock ssl.closelock begin
         # we do an additional check once inside the lock in case
         # it was closed while we were waiting on the lock
         isopen(ssl) || return
```

Comment:
Yeah, good call. I think we should change this to a `ssl.ssl.ssl == C_NULL && return` check instead. What we want to avoid is racing `close` calls. I guess we could do this with a CAS operation instead; i.e. if you win the CAS, then you get to call `free`. Maybe I'll change it to that.

Reply:
> CAS operation instead;

I think this is exactly what I propose in the second comment :+1: